### ISSUE TITLE
Pluses in parent user:pass auth

### ISSUE THREAD

**Issue by emarudee** [z3APA3A/3proxy#305]
1) example of config:
maxconn 10
auth strong
users user1:CL:pass1
allow user1 * * *
parent 1000 http 10.10.10.10 8000 "3569+US+87275089985758" qwerty123
proxy -4 -olSO_REUSEADDR,SO_REUSEPORT -ocTCP_TIMESTAMPS,TCP_NODELAY -osTCP_NODELAY -n -a -p16000
flush

2) this curl command works fine
curl --proxy "3569+US+87275089985758:qwerty123@10.10.10.10:8000" https://ipinfo.io/json

3) test command
curl --proxy "user1:pass1@127.0.0.1:16000" https://ipinfo.io/json
Throws 407 Authentication error
Which is parent proxy throws if user:pass auth is wrong.

How do I add correctly username with "+" symbols?

**Comment by emarudee** [comment]
The response is different.

1) if using correct user1:pass1
< HTTP/1.0 407 Unauthorized
< Cache-Control: no-cache
< Content-Type: text/html
< Connection: close
< 
* Received HTTP code 407 from proxy after CONNECT
* Closing connection 0

Its from parent proxy server

2) If using wrong user1:pass12341252151
< HTTP/1.0 407 Proxy Authentication Required
< Proxy-Authenticate: Basic realm="proxy"
* Authentication problem. Ignoring this.
< Connection: close
< Content-type: text/html; charset=utf-8
< 
* Received HTTP code 407 from proxy after CONNECT
* Closing connection 0

This is the response from 3proxy

**Comment by z3APA3A** [comment]
Try to use
parent 1000 http 10.10.10.10 8000 "3569 US 87275089985758" qwerty123
because in URI encoding "+" replaces space, so you login probably does not contain any "+" signs.

**Comment by emarudee** [comment]
It does not work. Still the same 407 error.
Even trying

curl --proxy "3569 US 87275089985758:qwerty123@10.10.10.10:8000" https://ipinfo.io/json
Also throws 407

**Comment by emarudee** [comment]
replacing + to %2B in curl command works
curl --proxy "3569%2BUS%2B87275089985758:qwerty123@10.10.10.10:8000" https://ipinfo.io/json

However in 3proxy it does not work.

**Comment by z3APA3A** [comment]
Add logging within 3proxy, check the error code, if it's really parent proxy issue check the traffic with tcpdump, compare credentials sent from cutl with ones sent with 3proxy.

**Comment by emarudee** [comment]
Thanks. It will take some time to do that.
I will let you know what I found there.

**Comment by z3APA3A** [comment]
There can be a problem e.g. if prarent proxy doesn't support basic authentication.

**Comment by emarudee** [comment]
Parent proxy support basic authentication.
Curl -v command shows 
Proxy auth using Basic with user ....
With 200 OK status which means its okay.

It will take several hours to tcpdump and find out auth header 3proxy sends to parent proxy.

**Comment by emarudee** [comment]
Are we able to manually set basic auth header for parent proxy?
Proxy-Authorization: Basic {BASE64_CODE}

**Comment by Niek** [comment]
Same problem here, is there a workaround?

**Comment by z3APA3A** [comment]
You can report it as a bug to parent proxy operators.

https://tools.ietf.org/html/rfc7617#page-3

> 
>    Note that both scheme and parameter names are matched case-
>    insensitively.

**Comment by Niek** [comment]
I agree, this is a bug in the parent proxy implementation. Still, I think it'd make more sense to use ``Basic`` instead of ``basic`` because as far as I can tell most user agents (curl, wget, Chrome, etc) use it that way.

**Comment by z3APA3A** [comment]
I have changed it in devel, thanks.

**Comment by raintears2** [comment]
Same problem here,I changing basic to Basic in the 2 files,but it still got "407 Proxy Authentication Required".  What's the 3proxy version?I used 3proxy 0.8.13. I checked the log of 3proxy,the error code is 00000 and I test curl the parent proxy work fine in the same server.

Below is my config section:

allow * MYUSERNAME

parent 1000 http zproxy.lum-superproxy.io 22225 lum-customer-hl_XXXX-zone-sharedip-ip-YY.YY.YY.YY PROXYPASSWORD

socks -a -p30001 -i x.x.x.x


-------------------------------------------------------------------
**Edit**: found the issue, the problem is not the pluses but the fact that 3proxy send the header ``Proxy-Authorization: basic <base64-encoded-auth>``, changing ``basic`` to ``Basic`` fixes the problem. Lines in question:
https://github.com/z3APA3A/3proxy/blob/22cf9254c57e6f9b55df768cd3800697ae4ad4a3/src/auth.c#L58
https://github.com/z3APA3A/3proxy/blob/22cf9254c57e6f9b55df768cd3800697ae4ad4a3/src/proxy.c#L862

_Originally posted by @Niek in https://github.com/z3APA3A/3proxy/issues/305#issuecomment-542570806_
